File: LeetCode/Medium/0022-generate-parentheses/0022-generate-parentheses.cs
public class Solution {
    public IList<string> GenerateParenthesis(int n) {
        var results = new List<string>();
        GenerateParenthesis(n, 0, 0, 0, new char[2 * n], results);
        return results;
    }

    private void GenerateParenthesis(int n, int opened, int closed, int idx, char[] result, IList<string> results)
    {
        if (idx >= 2 * n)
        {
            results.Add(new string(result));
            return;
        }

        if (opened < n)
        {
            result[idx] = '(';
            GenerateParenthesis(n, opened + 1, closed, idx + 1, result, results);
        }
        if (opened > closed)
        {
            result[idx] = ')';
            GenerateParenthesis(n, opened, closed + 1, idx + 1, result, results);
        }
    }
}

File: LeetCode/Medium/0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        stack: Deque[Tuple[char: str, opened: int, closed: int]]  = deque()
        stack.append(("(", 1, 0))
        parentheses_combinations = []
        combination_len = 2 * n
        combination = [" "] * combination_len

        while stack:
            parenthese, opened, closed = stack.pop()
            combination[opened + closed - 1] = parenthese
            if opened + closed == combination_len:
                parentheses_combinations.append("".join(combination))
            else:
                if opened < n:
                    stack.append(("(", opened + 1, closed))
                if closed < opened:
                    stack.append((")", opened, closed + 1))

        return parentheses_combinations
        
        